jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionData.java

```diff
@@ -78,22 +79,47 @@ public static void serializeAttributes (SessionData data, java.io.ObjectOutputSt
         out.writeObject(entries);
         for (Entry<String,Object> entry: data._attributes.entrySet())
         {
-            out.writeUTF(entry.getKey());     
-            ClassLoader loader = entry.getValue().getClass().getClassLoader();
-            boolean isServerLoader = false;
-
-            if (loader == Thread.currentThread().getContextClassLoader()) //is it the webapp classloader?
-                isServerLoader = false;
-            else if (loader == Thread.currentThread().getContextClassLoader().getParent() || loader == SessionData.class.getClassLoader() || loader == null) // is it the container loader?
-                isServerLoader = true;
-            else
-                throw new IOException ("Unknown loader"); // we don't know what loader to use
+            out.writeUTF(entry.getKey());
             
-            out.writeBoolean(isServerLoader);
+            Class<?> clazz = entry.getValue().getClass();
+            ClassLoader loader = clazz.getClassLoader();
+            ClassLoader contextLoader = Thread.currentThread().getContextClassLoader();
+            boolean isContextLoader;
+
+            if (loader == contextLoader) //is it the context classloader?
+                isContextLoader = true;
+            else if (contextLoader == null) //not context classloader
+                isContextLoader = false;
+            else if (contextLoader instanceof ClassVisibilityChecker)
+            {
+                //Clazz not loaded by context classloader, but ask if loadable by context classloader,
+                //because preferable to use context classloader if possible (eg for deep structures).
+                ClassVisibilityChecker  checker = (ClassVisibilityChecker)(contextLoader);
+                isContextLoader = (checker.isSystemClass(clazz) && !(checker.isServerClass(clazz)));
+            }
+            else
+            {
+                //Class wasn't loaded by context classloader, but try loading from context loader,
+                //because preferable to use context classloader if possible (eg for deep structures).
+                try 
+                { 
+                    Class<?> result = contextLoader.loadClass(clazz.getName());
```

Comment:
I think it would be wise to update https://www.eclipse.org/jetty/documentation/9.4.x/jetty-classloading.html#using-custom-webappclassloader specifically:

> We recommend, but do not require, that your custom loader subclasses WebAppClassLoader. 

I think that will need a note to the effect that if they do not subclass WebAppClassLoader and they are using session persistence then they probably will want their custom classloader to implement `ClassVisibilityChecker` or they may see a performance impact in session persistence.